pkg/controllers/jobset_controller.go

```diff
@@ -15,45 +15,263 @@ package controllers
+// a particular one, we won't requeue just to finish the deleting.
+func (r *JobSetReconciler) cleanUpOldJobs(ctx context.Context, jobs *childJobs) {
+	// Clean up failed jobs
+	for _, job := range jobs.failed {
+		if err := r.Delete(ctx, job, client.PropagationPolicy(metav1.DeletePropagationBackground)); client.IgnoreNotFound(err) != nil {
+			klog.Error(err, "unable to delete old failed job", "job", job)
+		} else {
+			klog.V(0).Info("deleted old failed job", "job", job)
+		}
+	}
+}
+
+// getChildJobs fetches all Jobs owned by the JobSet and returns them
+// categorized by status (active, successful, failed).
+func (r *JobSetReconciler) getChildJobs(ctx context.Context, jobSet *jobsetv1alpha.JobSet, req ctrl.Request) (*childJobs, error) {
+	// Get all active jobs owned by JobSet.
+	var childJobList batchv1.JobList
+	if err := r.List(ctx, &childJobList, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		klog.Error(err, "unable to list child Jobs")
+		return nil, err
+	}
+
+	jobs := childJobs{}
+	for i, job := range childJobList.Items {
+		_, finishedType := IsJobFinished(&job)
+		switch finishedType {
+		case "": // active
+			jobs.active = append(jobs.active, &childJobList.Items[i])
+		case batchv1.JobFailed:
+			jobs.failed = append(jobs.failed, &childJobList.Items[i])
+		case batchv1.JobComplete:
+			jobs.successful = append(jobs.successful, &childJobList.Items[i])
+		}
+	}
+
+	return &jobs, nil
+}
+
+func (r *JobSetReconciler) createReadyJobs(ctx context.Context, req ctrl.Request, jobSet *jobsetv1alpha.JobSet) error {
+	for _, jobTemplate := range jobSet.Spec.Jobs {
+		job, err := r.constructJobFromTemplate(jobSet, &jobTemplate)
+		if err != nil {
+			klog.Error(err, "unable to construct job from template", "jobTemplate", jobTemplate)
+			return err
+		}
+
+		// Skip the job if it is already active or succeeded.
+		skip, err := r.shouldSkipJob(job)
+		if err != nil {
+			return err
+		}
+		if skip {
+			klog.Infof("skipping job %s", job.Name)
+			continue
+		}
+
+		klog.Infof("creating job %s", job.Name)
+
+		// First create headless service if specified for this job.
+		if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+			if err := r.createHeadlessSvcIfNotExist(ctx, req, jobSet, job); err != nil {
+				return err
+			}
+			// Update job spec to set subdomain as headless service name (will always be same as job name)
+			job.Spec.Template.Spec.Subdomain = job.Name
+		}
+
+		if err := r.Create(ctx, job); err != nil {
+			klog.Error(err, "unable to create Job for JobSet", "job", job)
+			return err
+		}
+
+		klog.Info("created Job for JobSet run", "job", job)
+	}
+	return nil
+}
+
+func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, req ctrl.Request, jobSet *jobsetv1alpha.JobSet, job *batchv1.Job) error {
+	// Check if service already exists. Service name is same as job name.
```

Comment:
We need to think about adopting the service and updating the selector if it is not matching the job's selector. At least leave at a comment to that effect for now while we think about pros/cons and semantics for this case.

Reply:
Left a TODO for this